File: 02 - Breadth First Search/M616.CourseScheduleII.py
class Solution:
    """
    @param: numCourses: a total of n courses
    @param: prerequisites: a list of prerequisite pairs
    @return: the course order
    """
    def findOrder(self, numCourses, prerequisites):
        # write your code here
        if not prerequisites:
            return [num for num in range(numCourses)]

        courses, edges, degree = [], {num: [] for num in range(numCourses)}, [0] * numCourses
        for i, j in prerequisites:
            edges[j].append(i)
            degree[i] += 1

        queue = collections.deque([num for num in range(numCourses) if degree[num] == 0])
        while queue:
            prerequisite = queue.popleft()
            courses.append(prerequisite)

            for course in edges[prerequisite]:
                degree[course] -= 1
                if degree[course] == 0:
                    queue.append(course)

        if len(courses) == numCourses:
            return courses

        return []

File: 02 - Breadth First Search/M616.CourseScheduleII.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        // List[] edges = new ArrayList[numCourses];
        Map<Integer, List<Integer>> edges = new HashMap<>();
        int[] degree = new int[numCourses];

        // for (int i = 0; i < numCourses; i++) {
        //     edges[i] = new ArrayList<Integer>();
        // }

        for (int[] pair: prerequisites) {
            int dest = pair[0], src = pair[1];

            List<Integer> preList = edges.getOrDefault(src, new ArrayList<>());
            preList.add(dest);
            edges.put(src, preList);

            degree[dest]++;
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (degree[i] == 0) {
                q.offer(i);
            }
        }

        int[] result = new int[numCourses];
        int count = 0;
        while (!q.isEmpty()) {
            int course = q.poll();
            result[count++] = course;
            if (!edges.containsKey(course)) continue;

            for (Integer dest: edges.get(course)) {
                degree[dest]--;
                if (degree[dest] == 0) {
                    q.offer(dest);
                }
            }
        }

        if (count == numCourses) {
            return result;
        }

        return new int[0];
    }
}
